### PR TITLE
[mono] Add MonoAOTCompiler MSBuild Task to produce NuPkg

### PR DESCRIPTION
**Pull request by mdh1418**
Part of #46429

This PR produces a NuGet package containing the [MonoAOTCompiler task](https://github.com/dotnet/runtime/tree/master/src/tasks/AotCompilerTask) and imports the `MonoAOTCompiler` task for immediate use.

### CODE REVIEW

File: src/mono/netcore/nuget/Microsoft.NET.Runtime.MonoAOTCompiler.Task/Sdk/Sdk.props [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+          created a backup copy.  Incorrect changes to this file will make it
+          impossible to load or build your projects from the command-line or the IDE.
+
+Copyright (c) .NET Foundation. All rights reserved.
+***********************************************************************************************
+-->
+<Project ToolsVersion="14.0">
+
+  <PropertyGroup>
+    <_AOTTasksDirectoryRoot Condition=" '$(_ILLinkTasksDirectoryRoot)' == '' ">$(MSBuildThisFileDirectory)../tools/</_AOTTasksDirectoryRoot>
```
Comment (steveisok) [inline src/mono/netcore/nuget/Microsoft.NET.Runtime.MonoAOTCompiler.Task/Sdk/Sdk.props:15]: The condition should be `_AOTTasksDirectoryRoot`

Comment (mdh1418) [reply src/mono/netcore/nuget/Microsoft.NET.Runtime.MonoAOTCompiler.Task/Sdk/Sdk.props]: Thanks!

File: src/mono/netcore/nuget/Microsoft.NET.Runtime.MonoAOTCompiler.Task/Sdk/Sdk.props [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+    <_AOTTasksTFM>net5.0</_AOTTasksTFM>
```
Comment (steveisok) [inline src/mono/netcore/nuget/Microsoft.NET.Runtime.MonoAOTCompiler.Task/Sdk/Sdk.props:16]: I don't know if the TFM should be hardcoded or not.  It should follow whatever `$(NetCoreAppToolCurrent)` is, but it's not a value when a developer would use the package.  

If we leave it hard coded, then we would need to change it when `$(NetCoreAppToolCurrent)` bumps.  Chances are pretty good we would not remember and bad things will happen.

**Comment by vargaz** [comment]
Shouldn't all wasm build tasks be put into one nuget package ?

File: src/mono/netcore/nuget/Microsoft.NET.Runtime.MonoAOTCompiler.Task/Sdk/Sdk.props (hunk 2 above)
Comment (mdh1418) [reply src/mono/netcore/nuget/Microsoft.NET.Runtime.MonoAOTCompiler.Task/Sdk/Sdk.props:16]: Is it necessary to have this file? Could the user importing the package create their own UsingTask as how it is done here? https://github.com/dotnet/samples/blob/1bb5dc93e2ce9edafddbee8c9dfd25cd31dafdf1/core/mono-samples/wasm/WasmSampleApp.csproj#L11-L16

Comment (steveisok) [reply src/mono/netcore/nuget/Microsoft.NET.Runtime.MonoAOTCompiler.Task/Sdk/Sdk.props:16]: Good question.  My preference would be to have all you need in the package (keep it).  That way the xamarin sdk's and whoever else can just use it.

**Comment by mdh1418** [comment]
@vargaz I thought MonoAOTCompiler could apply to iOS and Android as well

**Comment by steveisok** [comment]
> Shouldn't all wasm build tasks be put into one nuget package ?

This package is intended to be used beyond runtime.  Each of the sdk's will define a workload that will contain a bunch of workload packs (the AOT MSBuild Task is one of them).  That will allow the developer to choose (or not) to install different parts of the dotnet sdk.

File: src/mono/netcore/nuget/Microsoft.NET.Runtime.MonoAOTCompiler.Task/Sdk/Sdk.props (hunk 2 above)
Comment (marek-safar) [reply src/mono/netcore/nuget/Microsoft.NET.Runtime.MonoAOTCompiler.Task/Sdk/Sdk.props:16]: Why is TFM needed here?

Comment (mdh1418) [reply src/mono/netcore/nuget/Microsoft.NET.Runtime.MonoAOTCompiler.Task/Sdk/Sdk.props:16]: `_AOTTasksTFM` is included to resemble the path to `MonoAOTCompiler.dll` in the NuPkg to make the task readily available to those using this NuPkg via `<UsingTask TaskName="MonoAOTCompiler" AssemblyFile="$(AOTTasksAssembly)" />` at the end of this file.

`MonoAOTCompiler.dll` is located at `tools\$(NetCoreAppToolCurrent)\` in the NuPkg by 
`<PackageFile Include="@(_AOTCompilerTaskFile)" TargetPath="tools\$(NetCoreAppToolCurrent)\" />` in `Microsoft.NET.Runtime.MonoAOTCompiler.Task.pkgproj`.

File: src/mono/netcore/nuget/Microsoft.NET.Runtime.MonoAOTCompiler.Task/Sdk/Sdk.props [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+Copyright (c) .NET Foundation. All rights reserved.
+***********************************************************************************************
+-->
+<Project ToolsVersion="14.0">
+
+  <PropertyGroup>
+    <_AOTTasksDirectoryRoot Condition="'$(_AOTTasksDirectoryRoot)' == ''">$(MSBuildThisFileDirectory)../tools/</_AOTTasksDirectoryRoot>
+    <_AOTTasksTFM>net5.0</_AOTTasksTFM>
+    <_AOTTasksDirectory>$(_AOTTasksDirectoryRoot)$(_AOTTasksTFM)/</_AOTTasksDirectory>
+    <AOTTasksAssembly Condition="'$(AOTTasksAssembly)' == ''">$(_AOTTasksDirectory)MonoAOTCompiler.dll</ILLinkTasksAssembly>
```
[the remark this answers is not in the document]
Comment (mdh1418) [reply src/mono/netcore/nuget/Microsoft.NET.Runtime.MonoAOTCompiler.Task/Sdk/Sdk.props]: Thanks, Removed `_AOTTasksTFM` and `_AOTTasksDirectory`. Kept `_AOTTasksDirectoryRoot` so the user can still control the value.

File: src/mono/netcore/nuget/Microsoft.NET.Runtime.MonoAOTCompiler.Task/Microsoft.NET.Runtime.MonoAOTCompiler.Task.pkgproj [hunk 4]
```diff
@@ -0,0 +1,21 @@
+<Project DefaultTargets="Build">
+  <Import Project="$([MSBuild]::GetPathOfFileAbove(Directory.Build.props))" />
+
+  <PropertyGroup>
+    <PackageDescription>The package that contains the MonoAOTCompiler MSBuild Task</PackageDescription>
```
Comment (radical) [inline src/mono/netcore/nuget/Microsoft.NET.Runtime.MonoAOTCompiler.Task/Microsoft.NET.Runtime.MonoAOTCompiler.Task.pkgproj:5]: `Provides the MonoAOTCompiler msbuild task`

File: src/mono/netcore/nuget/Microsoft.NET.Runtime.MonoAOTCompiler.Task/Sdk/Sdk.props [hunk 5]
```diff
@@ -0,0 +1,20 @@
+<!--
```
Comment (radical) [inline src/mono/netcore/nuget/Microsoft.NET.Runtime.MonoAOTCompiler.Task/Sdk/Sdk.props:1]: IIUC, this package is not meant to be used as a SDK. So, this file should instead be named `build/Microsoft.NET.Runtime.MonoAOTCompiler.Task.props`.

File: src/mono/netcore/nuget/Microsoft.NET.Runtime.MonoAOTCompiler.Task/Sdk/Sdk.props [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+          created a backup copy.  Incorrect changes to this file will make it
+          impossible to load or build your projects from the command-line or the IDE.
+
+Copyright (c) .NET Foundation. All rights reserved.
+***********************************************************************************************
+-->
+<Project ToolsVersion="14.0">
+
+  <PropertyGroup>
+    <_AOTTasksDirectoryRoot Condition="'$(_AOTTasksDirectoryRoot)' == ''">$(MSBuildThisFileDirectory)../tools/</_AOTTasksDirectoryRoot>
```
Comment (radical) [inline src/mono/netcore/nuget/Microsoft.NET.Runtime.MonoAOTCompiler.Task/Sdk/Sdk.props:15]: Why is this needed? This is part of the nuget, and should always be using the bundled task.

File: src/mono/netcore/nuget/Microsoft.NET.Runtime.MonoAOTCompiler.Task/Sdk/Sdk.props [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+    <AOTTasksAssembly Condition="'$(AOTTasksAssembly)' == ''">$(_AOTTasksDirectoryRoot)net5.0/MonoAOTCompiler.dll</ILLinkTasksAssembly>
```
Comment (radical) [inline src/mono/netcore/nuget/Microsoft.NET.Runtime.MonoAOTCompiler.Task/Sdk/Sdk.props:16]: This shouldn't be overridable either, IIUC.

File: src/mono/netcore/nuget/Microsoft.NET.Runtime.MonoAOTCompiler.Task/Sdk/Sdk.props [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+  </PropertyGroup>
+
+  <UsingTask TaskName="MonoAOTCompiler" AssemblyFile="$(AOTTasksAssembly)" />
```
Comment (radical) [inline src/mono/netcore/nuget/Microsoft.NET.Runtime.MonoAOTCompiler.Task/Sdk/Sdk.props:19]: This shouldn't be here. Instead the project using this package should have the `UsingTask`, or whatever is providing the targets to use this task.

File: src/tasks/AotCompilerTask/MonoAOTCompiler.csproj [hunk 9]
```diff
@@ -22,4 +22,6 @@
       <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
     </Content>
   </ItemGroup>
+
+  <Target Name="GetFilesToPackage" />
```
Comment (radical) [inline src/tasks/AotCompilerTask/MonoAOTCompiler.csproj:26]: Why is this needed? Comment would be useful

**Review by radical — CHANGES_REQUESTED** [review]
I think it would be very helpful to have an example project that uses this package. Or at least explain how this will get used. That would make it easier to think about where the `UsingTask` should be, for example.

File: src/tasks/AotCompilerTask/MonoAOTCompiler.csproj (hunk 9 above)
Comment (mdh1418) [reply src/tasks/AotCompilerTask/MonoAOTCompiler.csproj:26]: I think its required by the `ExpandProjectReferences` target in `Packaging.targets`, following the `Pack` target in `Sdk.Targets`. 

I had followed the convention that generated the `Microsoft.NETCore.BrowserDebugHost.Transport` NuPkg when creating the `Microsoft.NET.Runtime.Android.Sample.Mono`, `Microsoft.NET.Runtime.iOS.Sample.Mono`, and `Microsoft.NET.Runtime.wasm.Sample.Mono` NuPkgs, because otherwise `/Users/mdhwang/runtime_wasm_clean/src/tasks/AotCompilerTask/MonoAOTCompiler.csproj : error MSB4057: The target "GetFilesToPackage" does not exist in the project` would throw.

File: src/mono/netcore/nuget/Microsoft.NET.Runtime.MonoAOTCompiler.Task/Microsoft.NET.Runtime.MonoAOTCompiler.Task.pkgproj (hunk 4 above)
Comment (mdh1418) [reply src/mono/netcore/nuget/Microsoft.NET.Runtime.MonoAOTCompiler.Task/Microsoft.NET.Runtime.MonoAOTCompiler.Task.pkgproj:5]: Thanks :)

File: src/mono/netcore/nuget/Microsoft.NET.Runtime.MonoAOTCompiler.Task/Sdk/Sdk.props (hunk 8 above)
Comment (mdh1418) [reply src/mono/netcore/nuget/Microsoft.NET.Runtime.MonoAOTCompiler.Task/Sdk/Sdk.props:19]: Will remove the file and opt for the user themselves adding their own `<UsingTask` to their project.

File: src/mono/netcore/nuget/Microsoft.NET.Runtime.MonoAOTCompiler.Task/Sdk/Sdk.props (hunk 7 above)
Comment (mdh1418) [reply src/mono/netcore/nuget/Microsoft.NET.Runtime.MonoAOTCompiler.Task/Sdk/Sdk.props:16]: Removing the file as the NuPkg will no longer import the UsingTask.

File: src/mono/netcore/nuget/Microsoft.NET.Runtime.MonoAOTCompiler.Task/Sdk/Sdk.props (hunk 6 above)
Comment (mdh1418) [reply src/mono/netcore/nuget/Microsoft.NET.Runtime.MonoAOTCompiler.Task/Sdk/Sdk.props:15]: Removing the file as the NuPkg will no longer import the UsingTask.

File: src/mono/netcore/nuget/Microsoft.NET.Runtime.MonoAOTCompiler.Task/Sdk/Sdk.props (hunk 5 above)
Comment (mdh1418) [reply src/mono/netcore/nuget/Microsoft.NET.Runtime.MonoAOTCompiler.Task/Sdk/Sdk.props:1]: Removing the file as the NuPkg will no longer import the UsingTask.

Comment (radical) [reply src/mono/netcore/nuget/Microsoft.NET.Runtime.MonoAOTCompiler.Task/Sdk/Sdk.props:1]: It would still be useful to have `build/Microsoft.NET.Runtime.MonoAOTCompiler.Task.props` that populates a `$(MonoAOTCompilerTaskAssemblyPath)`, so that the project using this nuget can just use that, instead of having to construct the path. msbuild will automatically import this file.

Comment (mdh1418) [reply src/mono/netcore/nuget/Microsoft.NET.Runtime.MonoAOTCompiler.Task/Sdk/Sdk.props]: Something like
```
<PropertyGroup>
  <MonoAOTCompilerTaskAssemblyPath>tools\net5.0\MonoAOTCompiler.dll</MonoAOTCompilerTaskAssemblyPath>
</PropertyGroup>
```

Comment (radical) [reply src/mono/netcore/nuget/Microsoft.NET.Runtime.MonoAOTCompiler.Task/Sdk/Sdk.props]: `<MonoAOTCompilerTaskAssemblyPath>$(MSBuildThisFileDirectory)..\tools\net5.0\MonoAOTCompiler.dll</MonoAOTCompilerTaskAssemblyPath>`

Comment (mdh1418) [reply src/mono/netcore/nuget/Microsoft.NET.Runtime.MonoAOTCompiler.Task/Sdk/Sdk.props]: Thanks, done!

**Comment by mdh1418** [comment]
We use the `MonoAOTCompiler` task in the mono samples, just not as NuPkg imports. The main purpose of making the NuPkg is for external use, so those developing outside of `dotnet/runtime` can call the `MonoAOTCompiler` task. Such as how https://github.com/dotnet/samples/tree/master/core/mono-samples are using the respective NuPkgs. Where would be the best place to note that?

**Comment by radical** [comment]
> We use the `MonoAOTCompiler` task in the mono samples, just not as NuPkg imports. The main purpose of making the NuPkg is for external use, so those developing outside of `dotnet/runtime` can call the `MonoAOTCompiler` task. Such as how https://github.com/dotnet/samples/tree/master/core/mono-samples are using the respective NuPkgs. Where would be the best place to note that?

Maybe in a `README.md` next to the pkgproj.

Have you tried using the generated nuget with such a sample?

File: src/tasks/AotCompilerTask/MonoAOTCompiler.csproj (hunk 9 above)
Comment (radical) [reply src/tasks/AotCompilerTask/MonoAOTCompiler.csproj:26]: We could specify files to be packaged in the task project itself, like: https://github.com/radical/runtime/commit/b81753bbae2bd0f3c497e30054ae8af9e953daf5 .

File: src/mono/netcore/nuget/Microsoft.NET.Runtime.MonoAOTCompiler.Task/Sdk/Sdk.props (hunk 2 above)
Comment (mdh1418) [reply src/mono/netcore/nuget/Microsoft.NET.Runtime.MonoAOTCompiler.Task/Sdk/Sdk.props]: Instead, going to have the user import the NuPkg and standup the MonoAOTCompiler task with their own UsingTask, as detailed in the `README.md`

File: src/tasks/AotCompilerTask/MonoAOTCompiler.csproj (hunk 9 above)
Comment (mdh1418) [reply src/tasks/AotCompilerTask/MonoAOTCompiler.csproj]: Thanks! I didn't realize that could be done 😄

**Review by steveisok — APPROVED** [review]

File: src/tasks/AotCompilerTask/MonoAOTCompiler.csproj (hunk 9 above)
Comment (mdh1418) [reply src/tasks/AotCompilerTask/MonoAOTCompiler.csproj]: Added that commit and added a comment to explain the target.

File: src/mono/netcore/nuget/Microsoft.NET.Runtime.MonoAOTCompiler.Task/README.md [hunk 10]
```diff
@@ -0,0 +1,35 @@
+# MonoAOTCompiler MSBuild Task NuPkg
+The MonoAOTCompiler MSBuild task is useful outside the context of `dotnet/runtime`. The task is made available through a NuGet Package by packaging the `MonoAOTCompiler.dll` assembly produced from building `MonoAOTCompiler.csproj`. To use the task in a project, download the NuGet package through setting up the proper source and importing the package.
```
Comment (radical) [inline src/mono/netcore/nuget/Microsoft.NET.Runtime.MonoAOTCompiler.Task/README.md:2]:
```suggestion
The `MonoAOTCompiler` MSBuild task is also useful outside the context of `dotnet/runtime`. The task is made available through a NuGet Package containing the `MonoAOTCompiler.dll` assembly produced from building `MonoAOTCompiler.csproj`. To use the task in a project, reference the NuGet package, with the appropriate nuget source.
```

Comment (mdh1418) [reply src/mono/netcore/nuget/Microsoft.NET.Runtime.MonoAOTCompiler.Task/README.md]: Thanks!

**Review by radical — APPROVED** [review]
LGTM! 👍
